src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -78,21 +79,33 @@ internal class Http2FrameWriter
 
         _hpackEncoder = new DynamicHPackEncoder(serviceContext.ServerOptions.AllowResponseHeaderCompression);
 
-        // In practice, this is bounded by the number of concurrent streams allowed + whatever overflow we allow
-        _channel = Channel.CreateUnbounded<Http2OutputProducer>(new UnboundedChannelOptions()
+        // This is bounded by the maximum number of concurrent Http2Streams per Http2Connection.
+        // This isn't the same as SETTINGS_MAX_CONCURRENT_STREAMS, but typically double (with a floor of 100)
+        // which is the max number of Http2Streams that can end up in the Http2Connection._streams dictionary.
+        // 
+        // Setting a lower limit of SETTINGS_MAX_CONCURRENT_STREAMS might be sufficient because a stream shouldn't
+        // be rescheduling itself after being completed or canceled, but we're going with the more conservative limit
+        // in case there's some logic scheduling completed or canceled streams unnecessarily.
+        _channel = Channel.CreateBounded<Http2OutputProducer>(new BoundedChannelOptions(maxStreamsPerConnection)
         {
             AllowSynchronousContinuations = _scheduleInline,
             SingleReader = true
         });
 
-        _window = initialWindowSize;
+        _connectionWindow = initialConnectionWindowSize;
 
         _writeQueueTask = Task.Run(WriteToOutputPipe);
     }
 
     public void Schedule(Http2OutputProducer producer)
     {
-        _channel.Writer.TryWrite(producer);
+        if (!_channel.Writer.TryWrite(producer))
+        {
+            // It should not be possible to exceed the bound of the channel.
+            var ex = new ConnectionAbortedException("Http2FrameWriter._channel exceeded bounds.");
```

Comment:
I'm not a fan of including names of fields in error messages that are logged.

What about something like: `Connection exceeded the number of allowed output operations.`

Reply:
```suggestion
            var ex = new ConnectionAbortedException("HTTP/2 connection exceeded the output operations maximum queue size.");
```

I'm fine with this. My original proposal had `Environment.FailFast("Http2FrameWriter._channel exceeded bounds.")`, so this was a compromise. Hopefully no one ever sees this.

Reply:
No fail fast in the server side code please.

Reply:
Understood. `Environment.FailFast` would make for an easy DoS if it's actually hittable. I just want to make sure that this channel is **really** bounded and not something that can grow without limit in some edge case, because the OOM killer can also DoS you server if this is in fact unbounded.

Reply:
I don't mind bounding the channel but dropping the stream doesn't need to crash the server process.